File: BeatLooper/View Controllers/PlayerViewController.m
//
//  PlayerViewController.m
//  BeatLooper
//
//  Created by Isaak Meier on 5/6/21.
//

#import "PlayerViewController.h"
#import "BLPPlayer.h"
@import MediaPlayer;

@interface PlayerViewController ()

@property (weak, nonatomic) IBOutlet UIButton *playButton;
@property (weak, nonatomic) IBOutlet UIProgressView *songProgressBar;
@property (weak, nonatomic) IBOutlet UIButton *loopButton;
@property (weak, nonatomic) IBOutlet UILabel *songTitleLabel;
@property (weak, nonatomic) IBOutlet UILabel *playerStatusLabel;
@property (weak, nonatomic) IBOutlet UITableView *queueTableView;
@property (weak, nonatomic) IBOutlet UIButton *skipForwardButton;
@property (weak, nonatomic) IBOutlet UIButton *skipBackButton;
@property (weak, nonatomic) IBOutlet UIButton *removeButton; // also addSongButton
@property (weak, nonatomic) IBOutlet UISlider *songProgressSlider;

@property BOOL userIsHoldingSlider;
@property int sliderUpdatesToIgnoreCount;
@property NSArray *songsForPlayer; // need to delay playerModel init until viewDidLoad

@end

@implementation PlayerViewController

- (id)initWithSongs:(NSArray *)songs coordinator:coordinator {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    if ((self = [storyboard instantiateViewControllerWithIdentifier:@"PlayerViewController"])) {
        // assign properties
        _coordinator = coordinator;
        _songsForPlayer = songs;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    if (!self.playerModel) {
        NSLog(@"timing issue");
    }

    self.queueTableView.delegate = self.playerModel;
    self.queueTableView.dataSource = self.playerModel;
    [self.queueTableView setEditing:YES];
    [self.queueTableView setAllowsMultipleSelectionDuringEditing:YES];
}

- (void)viewWillDisappear:(BOOL)animated {
}

- (void)updateNowPlayingInfoCenterWithTitle:(NSString *)title {
    MPNowPlayingInfoCenter *infoCenter = [MPNowPlayingInfoCenter defaultCenter];
    if (title) {
        NSDictionary<NSString *, id> *nowPlayingInfo = @{MPMediaItemPropertyTitle : title};
        infoCenter.nowPlayingInfo = nowPlayingInfo;
    }
}

- (void)handleExistenceError {
    UIAlertController *alert = [UIAlertController
                                     alertControllerWithTitle:@"Error"
                                     message:@"Sorry, we lost this file. Shittisgone. Please delete and re-add."
                                     preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction* okButton = [UIAlertAction
                                    actionWithTitle:@"Haha, Ok"
                                    style:UIAlertActionStyleDefault
                                    handler:^(UIAlertAction * action) {
                                        //Handle your yes please button action here
                                        [[self navigationController] popViewControllerAnimated:YES];
                                        return;
                                    }];
    [alert addAction:okButton];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    BLPPlayerState state = self.playerModel.playerState;
    NSLog(@"view will appear");
    [self playerDidChangeState:state];
}


- (IBAction)playOrPauseSong:(id)sender {
    BOOL success = [self.playerModel togglePlayOrPause];
    if (success) {
        BLPPlayerState state = self.playerModel.playerState;
        [self updatePlayButtonFromState:state];
    } else {
        NSLog(@"Play/pause failed on empty player");
    }
}

- (void)updatePlayButtonFromState:(BLPPlayerState)state {
    if (state == BLPPlayerSongPaused || state == BLPPlayerLoopPaused) {
        [self animateButtonToPlayIcon:YES];
    } else if (state == BLPPlayerSongPlaying || state == BLPPlayerLoopPlaying) {
        [self animateButtonToPlayIcon:NO];
    }
}

- (void)updateButtonsWithState:(BLPPlayerState)state {
    if (state == BLPPlayerLoopPlaying || state == BLPPlayerLoopPaused) {
        [self.skipForwardButton setHidden:YES];
        [self.removeButton setHidden:YES];
        [self.queueTableView setEditing:NO];
    } else {
        [self.skipForwardButton setHidden:NO];
        [self.removeButton setHidden:NO];
        [self.queueTableView setEditing:YES];
    }
}

- (void)updateSongSubtitleWithState:(BLPPlayerState)state {
    switch (state) {
        case BLPPlayerSongPlaying:
            [self.playerStatusLabel setText:@"Now Playing"];
            return;
        case BLPPlayerSongPaused:
            [self.playerStatusLabel setText:@"Song Paused"];
            return;
        case BLPPlayerLoopPaused:
            [self.playerStatusLabel setText:@"Loop Paused"];
            return;
        case BLPPlayerLoopPlaying:
            [self.playerStatusLabel setText:@"Now Looping"];
            return;
        case BLPPlayerEmpty:
            [self.playerStatusLabel setText:@"Just chillin'"];
            return;
    }
}

- (IBAction)skipBackButtonTapped:(id)sender {
    BOOL success = [self.playerModel skipBackward]; if (!success) {
        NSLog(@"Skipping backward failed");
    }
}

- (IBAction)skipForwardButtonTapped:(id)sender {
    BOOL success = [self.playerModel skipForward];
    if (!success) {
        NSLog(@"Skipping forward failed");
    }
    [self.queueTableView reloadData];
}

- (IBAction)loopButtonTapped:(id)sender {
    if (self.playerModel.playerState != BLPPlayerEmpty) {
        Beat *songToLoop = [[BLPBeatModel new] getSongFromSongName:self.songTitleLabel.text];
        if (songToLoop) {
            BOOL playerIsLooping = self.playerModel.playerState == BLPPlayerLoopPlaying
            || self.playerModel.playerState == BLPPlayerLoopPaused;
            [self.coordinator openLooperViewForSong:songToLoop
                                          isLooping:playerIsLooping];
        } else {
            NSLog(@"ok");
        }
    } else {
        NSLog(@"Player state empty");
    }
}

- (IBAction)removeButtonTapped:(id)sender {
    if ([self.removeButton.currentTitle isEqualToString:@"Remove"]) {
        [self.queueTableView reloadData];
        [self.removeButton setTitle:@"Add Songs" forState:UIControlStateNormal];
        [self.playerModel removeSelectedSongs];
    } else {
        [self.coordinator showAddSongsView];
    }
}
- (IBAction)songSliderDidTouchDown:(id)sender {
    self.userIsHoldingSlider = YES;
}

- (IBAction)songSliderWasReleased:(id)sender {
    if (self.playerModel.playerState == BLPPlayerEmpty) {
        [self.playerStatusLabel setText:@"Just chillin' ;)"];
    } else {
        [self.playerModel seekToProgressValue:self.songProgressSlider.value];
        self.sliderUpdatesToIgnoreCount = 5;
    }
    self.userIsHoldingSlider = NO;
}

- (void)startLoopWithTimeRange:(CMTimeRange)timeRange {
    BOOL success = [self.playerModel startLoopingTimeRange:timeRange];
    if (!success) {
        NSLog(@"Loop failed.");
        [self handleErrorStartingLoop];
    }
}

- (void)handleErrorStartingLoop {
    UIAlertController *alert = [UIAlertController
                                     alertControllerWithTitle:@"Ay va voi"
                                     message:@"Hey Buddy, we couldn't start the loop for some reason. If I had to make an educated guess, it's because you tried to loop what couldn't be looped. That is, the song probably couldn't be looped between the bars that you provided."
                                     preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction* okButton = [UIAlertAction
                                    actionWithTitle:@"Haha, Ok"
                                    style:UIAlertActionStyleDefault
                                    handler:^(UIAlertAction * action) {
                                        //Handle your yes please button action here
                                        return;
                                    }];
    [alert addAction:okButton];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)stopLooping {
    [self.playerModel stopLooping];
}

- (void)changeCurrentSongTo:(Beat *)newSong {
    if (self.playerModel.currentSong != newSong.title) {
        [self.playerModel changeCurrentSongTo:newSong];
    }
}

- (void)addSongToQueue:(Beat *)song {
    [self.playerModel addSongToQueue:song];
    [self.queueTableView reloadData];
}

- (void)setupProgressBar {
    NSLog(@"Setting up progress bar");
    NSProgress *progress = [self.playerModel getProgressForCurrentItem];
    [self.songProgressBar setObservedProgress:progress];
    [self.songProgressSlider setValue:0];
}


- (void)animateButtonToPlayIcon:(BOOL)shouldAnimateToPlayIcon {
    [UIView transitionWithView:self.playButton
                      duration:0.1
                       options:UIViewAnimationOptionTransitionCrossDissolve
                    animations:^{
        if (shouldAnimateToPlayIcon) {
            [self.playButton setImage:[UIImage imageNamed:@"icons8-play-button-100"] forState:UIControlStateNormal];
        } else {
            [self.playButton setImage:[UIImage imageNamed:@"icons8-pause-button-100"] forState:UIControlStateNormal];
        }
    }
    completion:nil];

}

#pragma mark BLPPlayerDelegate
- (void)playerDidChangeSongTitle:(NSString *)songTitle {
    [self.songTitleLabel setText:songTitle];
    [self updateNowPlayingInfoCenterWithTitle:songTitle];
}

- (void)playerDidChangeState:(BLPPlayerState)state {
    [self updatePlayButtonFromState:state];
    [self updateSongSubtitleWithState:state];
    [self updateButtonsWithState:state];
    
    if (state == BLPPlayerEmpty) {
        // please kill me
        [self.coordinator playerViewControllerRequestsDeath];
    }
}

- (void)currentItemDidChangeStatus:(AVPlayerItemStatus)status {
        switch (status) {
            case AVPlayerItemStatusReadyToPlay:
                // Ready to Play
                NSLog(@"Item ready to play");
                [self setupProgressBar];
                break;
            // TODO handle these cases for the user
            // the songs usually load instantly but there could be a problem
            case AVPlayerItemStatusFailed:
                // Failed. Examine AVPlayerItem.error
                [self.playerStatusLabel setText:@"Failed to load song. Please delete & re-add."];
                NSLog(@"Failed. Examine AVPlayerItem.error");
                break;
            case AVPlayerItemStatusUnknown:
                // Not ready
                NSLog(@"Not ready");
                break;
        }
}

- (void)didUpdateCurrentProgressTo:(double)fractionCompleted {
    if (!self.userIsHoldingSlider) {
        [self.songProgressBar setProgress:fractionCompleted];
        // we ignore a few updates when we seek to a new time because
        // theres a gross visual glitch otherwise
        if (self.sliderUpdatesToIgnoreCount == 0) {
            [self.songProgressSlider setValue:fractionCompleted];
        } else {
            self.sliderUpdatesToIgnoreCount--;
        }
    }
}

// also updates progress bar
- (void)requestTableViewUpdate {
    [self.queueTableView reloadData];
}

- (void)requestProgressBarUpdate {
    [self.songProgressSlider setValue:0.0];
    [self.songProgressBar setProgress:0.0];
    [self setupProgressBar];
}

- (void)selectedIndexesChanged:(NSUInteger)count {
    if (count == 0) {
        [self.removeButton setTitle:@"Add Songs" forState:UIControlStateNormal];
    } else {
        [self.removeButton setTitle:@"Remove" forState:UIControlStateNormal];
    }
}


@end

File: BeatLooper/View Controllers/PlayerViewController.swift
//
//  PlayerViewController.swift
//  BeatLooper 2
//
//  Created by Isaak Meier on 5/6/21.
//  Migrated to Swift
//

import UIKit
import CoreData
import CoreMedia
import MediaPlayer
import MediaPlayer
import CoreMedia

class PlayerViewController: UIViewController {
    
    // MARK: - IBOutlets
    @IBOutlet weak var playButton: UIButton!
    @IBOutlet weak var songProgressBar: UIProgressView!
    @IBOutlet weak var loopButton: UIButton!
    @IBOutlet weak var songTitleLabel: UILabel!
    @IBOutlet weak var playerStatusLabel: UILabel!
    @IBOutlet weak var queueTableView: UITableView!
    @IBOutlet weak var skipForwardButton: UIButton!
    @IBOutlet weak var skipBackButton: UIButton!
    @IBOutlet weak var removeButton: UIButton!
    @IBOutlet weak var songProgressSlider: UISlider!
    
    // MARK: - Properties
    weak var coordinator: Coordinator?
    private var userIsHoldingSlider: Bool = false
    private var sliderUpdatesToIgnoreCount: Int = 0
    private var playerModel: Player?
    
    // MARK: - Initialization
    init(coordinator: Coordinator) {
        super.init(nibName: nil, bundle: nil)
        self.coordinator = coordinator
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    // MARK: - Lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        Task {
            if let state = await playerModel?.playerState {
                playerDidChangeState(state)
            }
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        loopButton.imageView?.contentMode = .scaleAspectFit
    }
    
    // MARK: - Setup
    private func setupUI() {
        queueTableView.delegate = playerModel
        queueTableView.dataSource = playerModel
        queueTableView.isEditing = true
        queueTableView.allowsMultipleSelectionDuringEditing = true
    }
    
    // MARK: - Public Methods
    func setup(player: Player) async {
        self.playerModel = player
        _ = await player.togglePlayOrPause()
        let state = await player.playerState
        playerDidChangeState(state)
    }
    
    func startLoopWithTimeRange(_ timeRange: CMTimeRange) async {
        guard let player = playerModel else { return }
        
        let success = await player.startLoopingTimeRange(timeRange)
        if !success {
            print("Loop failed.")
            handleErrorStartingLoop()
        }
    }
    
    func stopLooping() async {
        _ = await playerModel?.stopLooping()
    }
    
    func changeCurrentSongTo(_ newSong: Beat) async {
        guard let player = playerModel else { return }
        
        let currentSong = await player.currentSong
        if currentSong != newSong.title {
            _ = await player.changeCurrentSongTo(newSong)
        }
    }
    
    func addSongToQueue(_ song: Beat) async {
        _ = await playerModel?.addSongToQueue(song)
        queueTableView.reloadData()
    }
    
    // MARK: - Private Methods
    private func updateNowPlayingInfoCenterWithTitle(_ title: String?) {
        let infoCenter = MPNowPlayingInfoCenter.default()
        if let title = title {
            let nowPlayingInfo: [String: Any] = [MPMediaItemPropertyTitle: title]
            infoCenter.nowPlayingInfo = nowPlayingInfo
        }
    }
    
    private func handleExistenceError() {
        let alert = UIAlertController(
            title: "Error",
            message: "Sorry, we lost this file. Shittisgone. Please delete and re-add.",
            preferredStyle: .alert
        )
        
        let okAction = UIAlertAction(title: "Haha, Ok", style: .default) { _ in
            self.navigationController?.popViewController(animated: true)
        }
        alert.addAction(okAction)
        present(alert, animated: true)
    }
    
    private func updatePlayButtonFromState(_ state: PlayerState) {
        switch state {
        case .songPaused, .loopPaused:
            animateButtonToPlayIcon(true)
        case .songPlaying, .loopPlaying:
            animateButtonToPlayIcon(false)
        case .empty:
            break
        }
    }
    
    private func updateButtonsWithState(_ state: PlayerState) {
        switch state {
        case .loopPlaying, .loopPaused:
            skipForwardButton.isHidden = true
            removeButton.isHidden = true
            queueTableView.isEditing = false
        default:
            skipForwardButton.isHidden = false
            removeButton.isHidden = false
            queueTableView.isEditing = true
        }
    }
    
    private func updateSongSubtitleWithState(_ state: PlayerState) {
        switch state {
        case .songPlaying:
            playerStatusLabel.text = "Now Playing"
        case .songPaused:
            playerStatusLabel.text = "Song Paused"
        case .loopPaused:
            playerStatusLabel.text = "Loop Paused"
        case .loopPlaying:
            playerStatusLabel.text = "Now Looping"
        case .empty:
            playerStatusLabel.text = "Just chillin'"
        }
    }
    
    private func setupProgressBar() async {
        print("Setting up progress bar")
        if let progress = await playerModel?.getProgressForCurrentItem() {
            songProgressBar.observedProgress = progress
        }
        songProgressSlider.value = 0
    }
    
    private func animateButtonToPlayIcon(_ shouldAnimateToPlayIcon: Bool) {
        UIView.transition(with: playButton, duration: 0.1, options: .transitionCrossDissolve) {
            if shouldAnimateToPlayIcon {
                self.playButton.setImage(UIImage(named: "icons8-play-button-100"), for: .normal)
            } else {
                self.playButton.setImage(UIImage(named: "icons8-pause-button-100"), for: .normal)
            }
        }
    }
    
    private func handleErrorStartingLoop() {
        let alert = UIAlertController(
            title: "Ay va voi",
            message: "Hey Buddy, we couldn't start the loop. The song probably couldn't be looped between the bars that you provided.",
            preferredStyle: .alert
        )
        
        let okAction = UIAlertAction(title: "Haha, Ok", style: .default)
        alert.addAction(okAction)
        present(alert, animated: true)
    }
    
    // MARK: - Actions
    @IBAction func playOrPauseSong(_ sender: UIButton) {
        guard let player = playerModel else { return }
        
        Task {
            let success = await player.togglePlayOrPause()
            if success {
                let state = await player.playerState
                await MainActor.run {
                    updatePlayButtonFromState(state)
                }
            } else {
                print("Play/pause failed on empty player")
            }
        }
    }
    
    @IBAction func skipBackButtonTapped(_ sender: UIButton) {
        guard let player = playerModel else { return }
        
        Task {
            let success = await player.skipBackward()
            if !success {
                print("Skipping backward failed")
            }
        }
    }
    
    @IBAction func skipForwardButtonTapped(_ sender: UIButton) {
        guard let player = playerModel else { return }
        
        Task {
            let success = await player.skipForward()
            if !success {
                print("Skipping forward failed")
            }
            await MainActor.run {
                queueTableView.reloadData()
            }
        }
    }
    
    @IBAction func loopButtonTapped(_ sender: UIButton) {
        guard let player = playerModel else {
            print("Player state empty")
            return
        }
        
        Task {
            let state = await player.playerState
            guard state != .empty else {
                print("Player state empty")
                return
            }
            
            let model = BeatModel()
            if let songToLoop = model.getSongFromSongName(songTitleLabel.text ?? "") {
                let playerIsLooping = state == .loopPlaying || state == .loopPaused
                await MainActor.run {
                    coordinator?.openLooperViewForSong(songToLoop, isLooping: playerIsLooping)
                }
            } else {
                print("ok")
            }
        }
    }
    
    @IBAction func removeButtonTapped(_ sender: UIButton) {
        if removeButton.currentTitle == "Remove" {
            queueTableView.reloadData()
            removeButton.setTitle("Add Songs", for: .normal)
            Task {
                await playerModel?.removeSelectedSongs()
            }
        } else {
            coordinator?.showAddSongsView()
        }
    }
    
    @IBAction func songSliderDidTouchDown(_ sender: UISlider) {
        userIsHoldingSlider = true
    }
    
    @IBAction func songSliderWasReleased(_ sender: UISlider) {
        Task {
            let state = await playerModel?.playerState
            if state == .empty {
                await MainActor.run {
                    playerStatusLabel.text = "Just chillin' ;)"
                }
            } else {
                _ = await playerModel?.seekToProgressValue(songProgressSlider.value)
                await MainActor.run {
                    sliderUpdatesToIgnoreCount = 5
                }
            }
            await MainActor.run {
                userIsHoldingSlider = false
            }
        }
    }
}

// MARK: - PlayerDelegate
extension PlayerViewController: PlayerDelegate {
    
    nonisolated func playerDidChangeSongTitle(_ songTitle: String) {
        Task { @MainActor in
            songTitleLabel.text = songTitle
            updateNowPlayingInfoCenterWithTitle(songTitle)
        }
    }
    
    nonisolated func playerDidChangeState(_ state: PlayerState) {
        Task { @MainActor in
            updatePlayButtonFromState(state)
            updateSongSubtitleWithState(state)
            updateButtonsWithState(state)
            
            if state == .empty {
                // Please kill me
                coordinator?.playerViewControllerRequestsDeath()
            }
        }
    }
    
    nonisolated func currentItemDidChangeStatus(_ status: AVPlayerItem.Status) {
        Task { @MainActor in
            switch status {
            case .readyToPlay:
                print("Item ready to play")
                await setupProgressBar()
            case .failed:
                playerStatusLabel.text = "Failed to load song. Please delete & re-add."
                print("Failed. Examine AVPlayerItem.error")
            case .unknown:
                print("Not ready")
            @unknown default:
                break
            }
        }
    }
    
    nonisolated func didUpdateCurrentProgressTo(_ fractionCompleted: Double) {
        Task { @MainActor in
            if !userIsHoldingSlider {
                songProgressBar.progress = Float(fractionCompleted)
                
                // We ignore a few updates when we seek to a new time because
                // there's a gross visual glitch otherwise
                if sliderUpdatesToIgnoreCount == 0 {
                    songProgressSlider.value = Float(fractionCompleted)
                } else {
                    sliderUpdatesToIgnoreCount -= 1
                }
            }
        }
    }
    
    nonisolated func requestTableViewUpdate() {
        Task { @MainActor in
            queueTableView.reloadData()
        }
    }
    
    nonisolated func requestProgressBarUpdate() {
        Task { @MainActor in
            songProgressSlider.value = 0.0
            songProgressBar.progress = 0.0
            await setupProgressBar()
        }
    }
    
    nonisolated func selectedIndexesChanged(_ count: Int) {
        Task { @MainActor in
            if count == 0 {
                removeButton.setTitle("Add Songs", for: .normal)
            } else {
                removeButton.setTitle("Remove", for: .normal)
            }
        }
    }
}
